webrender/src/clip_scroll_tree.rs

```diff
@@ -296,23 +299,41 @@ impl ClipScrollTree {
         let root_reference_frame_id = self.root_reference_frame_id();
         let root_viewport = self.nodes[&root_reference_frame_id].local_viewport_rect;
         self.update_node_transform(root_reference_frame_id,
-                                    &ScrollToWorldTransform::identity(),
-                                    &as_scroll_parent_rect(&root_viewport));
+                                   &LayerToWorldTransform::identity(),
+                                   &as_scroll_parent_rect(&root_viewport),
+                                   LayerPoint::zero());
     }
 
     fn update_node_transform(&mut self,
                              layer_id: ScrollLayerId,
-                             parent_world_transform: &ScrollToWorldTransform,
-                             parent_viewport_rect: &ScrollLayerRect) {
+                             parent_reference_frame_transform: &LayerToWorldTransform,
+                             parent_viewport_rect: &ScrollLayerRect,
+                             parent_accumulated_scroll_offset: LayerPoint) {
         // TODO(gw): This is an ugly borrow check workaround to clone these.
         //           Restructure this to avoid the clones!
-        let (node_transform_for_children, viewport_rect, node_children) = {
+        let (reference_frame_transform, viewport_rect, accumulated_scroll_offset, node_children) = {
             match self.nodes.get_mut(&layer_id) {
                 Some(node) => {
-                    node.update_transform(parent_world_transform, parent_viewport_rect);
-
-                    (node.world_content_transform.with_source::<ScrollLayerPixel>(),
+                    node.update_transform(parent_reference_frame_transform,
+                                          parent_viewport_rect,
+                                          parent_accumulated_scroll_offset);
+
+                    // The transformation we are passing is the transformation of the parent
+                    // reference frame and the offset is the accumulated offset of all the nodes
+                    // between us and the parent reference frame. If we are a reference frame,
+                    // we need to reset both these values.
+                    let (transform, offset) = match node.node_type {
+                        NodeType::ReferenceFrame(..) =>
+                            (node.world_viewport_transform, LayerPoint::zero()),
```

Comment:
The parent _accumulated_scroll_offset is rolled into the transformation of reference frames. Reference frames "reset" the offset because they establish a new coordinate system and origin in the tree.